lib/clerk/sdk.rb

```diff
@@ -106,5 +110,40 @@ def sms_messages
     def users
       Resources::Users.new(self)
     end
+
+    # Returns the decoded JWT payload without verifying if the signature is
+    # valid.
+    #
+    # WARNING: This will not verify whether the signature is valid. You
+    # should not use this for untrusted messages! You most likely want to use
+    # verify_token.
+    def decode_token(token)
+      JWT.decode(token, nil, false).first
+    end
+
+    # Decode the JWT and verify it's valid (verify claims, signature etc.)
+    #
+    # JWKS are cached for JWKS_CACHE_LIFETIME seconds, in order to avoid
+    # unecessary roundtrips. In order to invalidate the cache, pass
+    # `force_refresh_jwks: true`.
+    def verify_token(token, force_refresh_jwks: false)
+      unverified_claims = decode_token(token)
+
+      jwk_loader = ->(options) do
+        @cached_jwks = nil if options[:invalidate] || force_refresh_jwks
+        @cached_jwks = nil if @last_fetched_at && Time.now.to_i - @last_fetched_at > JWKS_CACHE_LIFETIME
+
+        @cached_jwks ||= begin
+          @last_fetched_at = Time.now.to_i
+          conn = Faraday::Connection.new(unverified_claims["iss"])
```

Comment:
🔧 We should add HTTP net timeouts to max 5 seconds.

Reply:
Actually reworked the PR to re-use the parent Faraday connection.

Currently we don't set any timeouts in the SDK, though it is possible for the user to set them. That said, the underlying adapter (e.g. `Net::HTTP` sets defaults to 30 secs.)

IMHO timeouts should be left as a configuration to the user, since they may vary depending on their location and connection. However, if we decide to set them, we should do so for all the SDK requests, not just this one.

WDYT?

Reply:
I agree but the default set to 30 is a lot for the JWKS request.

Reply:
Cool 👍🏽 I've set it to 5 seconds.